src/federation.rs

```diff
@@ -0,0 +1,268 @@
+    /// block. Then from the next block which height is 100, Tapyrus network would get started to
+    /// use new aggreted public key to verify blocks.
+    block_height: u64,
+    /// The theshold which is requirement number of signer's agreements to produce block signatures.
+    threshold: u8,
+    /// Verifiable Secre Share and commitments from all signers in the federation.
+    nodevss: Vec<Vss>,
+}
+
+impl Federation {
+    pub fn new(public_key: PublicKey, block_height: u64, threshold: u8, nodevss: Vec<Vss>) -> Self {
+        Self {
+            signer_id: SignerID::new(public_key),
+            block_height,
+            threshold,
+            nodevss,
+        }
+    }
+
+    pub fn node_index(&self) -> usize {
+        self.signers()
+            .iter()
+            .position(|i| *i == self.signer_id)
+            .unwrap()
+    }
+
+    pub fn signers(&self) -> Vec<SignerID> {
+        let mut signers: Vec<SignerID> = self
+            .nodevss
+            .iter()
+            .map(|i| SignerID::new(i.sender_public_key.clone()))
+            .collect();
+        signers.sort();
+        signers
+    }
+
+    pub fn block_height(&self) -> u64 {
+        self.block_height
+    }
+    pub fn threshold(&self) -> u8 {
+        self.threshold
+    }
+    pub fn nodevss(&self) -> &Vec<Vss> {
+        &self.nodevss
+    }
+
+    /// Returns Map collection of received shares from all each signers in Key Generation Protocol
+    pub fn node_shared_secrets(&self) -> SharedSecretMap {
+        let mut secret_shares = SharedSecretMap::new();
+        for vss in &self.nodevss {
+            secret_shares.insert(
+                SignerID {
+                    pubkey: vss.sender_public_key,
+                },
+                SharedSecret {
+                    vss: VerifiableSS {
+                        parameters: ShamirSecretSharing {
+                            threshold: (self.threshold - 1) as usize,
+                            share_count: self.nodevss.len(),
+                        },
+                        commitments: vss
+                            .positive_commitments
+                            .iter()
+                            .map(|i| i.to_point())
+                            .collect(),
+                    },
+                    secret_share: vss.positive_secret,
+                },
+            );
+        }
+        secret_shares
+    }
+
+    /// Returns an aggregated share of the node.
+    pub fn node_secret_share(&self) -> SharedKeys {
+        let secret_shares = self.node_shared_secrets();
+
+        let shared_keys =
+            Sign::verify_vss_and_construct_key(&secret_shares, &(self.node_index() + 1))
+                .expect("invalid vss");
+        shared_keys
+    }
+
+    pub fn validate(&self) -> Result<(), Error> {
+        // Check all sender is different.
+        let signers = self.signers();
+        let is_overlap = signers
```

Comment:
This check is redundant because a double loop is contained.
Consider using HashSet.
```
let unique_set: HashSet<SignerID> = signers.into_iter().collect();
let is_overlap = unique_set.len() == signers.len();
```